src/api/callback.cc

```diff
@@ -271,6 +307,19 @@ MaybeLocal<Value> MakeCallback(Isolate* isolate,
                                int argc,
                                Local<Value> argv[],
                                async_context asyncContext) {
+#if defined(NODE_USE_NATIVE_ALS) && NODE_USE_NATIVE_ALS
+  return InternalMakeCallback(
+      isolate, recv, callback, argc, argv, asyncContext, Undefined(isolate));
```

Comment:
NAPI has modifications for that too. I was trying to avoid changing existing `MakeCallback` signatures by using scopes instead, but there's still the problem of storing the data on class instances, so I guess more work would need to be done to make it ABI compatible. 🤔 

I believe that the way this was written it would also exclude userland modules from getting the correct context frame with MakeCallback. I was also originally looking at modifying the `async_context` type directly which would have enabled that but was even more clearly an ABI break. I suppose adding a destructor to `async_context` would _also_ be a breaking change? 🤔

Reply:
Not sure if adding a destructor to `async_context` is breaking. Maybe it's only semver minor. But I guess the impact would be significant because quite some npm modules with native addons are distributed with prebuild binaries. So either module authors stick on an old node version to build these binaries or consumers have to use a newer node.js version.

Got the connection to `v8Impl::AsyncContext` now. Advantage here is that this one is internal. But not sure if it is really connected correct to the remaining infrastructure. It picks up current context on creation which looks good. But e.g. in `CallbackScope` only the resource and a asyncIds are forwarded to `node::CallbackScope` but not the context frame.

Quite some infrastructure is still based on asyncIds/async_wrap/async_resource. Not sure if we can get rid of this/extend it in a non breaking way or by adding new APIs. There are already now slight differences between async hooks based context propagation and async resource base context propagation (see https://github.com/nodejs/node/issues/43038) which can't be fixed without breaking changes in `MakeCallback` area to my understanding. I assume by moving to async context frame (and maybe later to native AsyncContext) we will get again some new differences. Hope they are acceptable as `AsyncLocalStore` is marked as stable.

Reply:
The intent is to land behind a runtime flag, so breaking behaviour is probably acceptable at least until it would be unflagged. My aim is for this to be as near as possible a drop-in replacement though, so ideally there's not a bunch of edge cases where it doesn't work the same.